### ISSUE TITLE
Inherited __LINUX__ value issue due to incorrect usage of add_definitions() in ./coreCLR/CMakeLists.txt

### ISSUE THREAD

**Issue by leemgs** [dotnet/coreclr#3504]
The add_definitions() is intended to add preprocessor definitions. We may use set() command to set a CMake, cache or environment variable to a given value. Below is my test result.  From our evaluation, the below patch will help inheritance issue in case that we try to define value on top folder.  

###  test environment:
- host pc: Ubuntu 14.04 x86 64bit PC
- command: time ROOTFS_DIR=/work/dotnet/rootfs-coreclr/arm  \
            ./build.sh arm debug clean  cross skipmscorlib | grep INFO
- commit no. : commit 50d84032861511d703125d80b9b4a6da11c2650a (Date:   Mon Feb 22 13:47:27 2016 -0800)

    Merge pull request #3293 from jkotas/r2r-default-callconv
    
    Use default calling convention small type widening rules for R2R

- related PR: https://github.com/dotnet/coreclr/pull/3479

### related patch in my local branch:

```

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 825dae4..1990fb5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -293,7 +293,8 @@ if (CLR_CMAKE_PLATFORM_UNIX)
   add_definitions(-DPLATFORM_UNIX=1)
 
   if(CLR_CMAKE_PLATFORM_LINUX)
-    add_definitions(-D__LINUX__=1)
+    add_definitions(-D__LINUX__)
+    set(__LINUX__ 1)
     if(CLR_CMAKE_PLATFORM_UNIX_TARGET_AMD64)
       message("Detected Linux x86_64")
       add_definitions(-DLINUX64)
```

###  evaluation: 

before) INFO-pal2: Configuring for [Linux], and [ ]
```

geunsik@FCS-Server1:~/dotnet/coreclr$ vi ./CMakeLists.txt 
geunsik@FCS-Server1:~/dotnet/coreclr$ time ROOTFS_DIR=/work/dotnet/rootfs-coreclr/arm  ./build.sh arm debug clean  cross skipmscorlib | grep INFO
loading initial cache file /home/geunsik/dotnet/coreclr/cross/arm/tryrun.cmake
Detected Linux ARM
-- INFO-pal2: Detected Linux (from ./src/pal/src/configure.cmake) 
-- INFO-pal2: Configuring for [Linux], and [ ]
CMake Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:
   HAVE_CLOCK_MONOTONIC_COARSE_EXITCODE (advanced)
For details see /home/geunsik/dotnet/coreclr/bin/obj/Linux.arm.Debug/TryRunResults.cmake
-- INFO-pal1: Detected Linux (from src/pal/src/CMakeLists.txt)
-- INFO-pal1: Configuring for [Linux], and []

real	0m7.328s
user	0m5.246s
sys	0m2.136s
geunsik@FCS-Server1:~/dotnet/coreclr$ 
```

after)  INFO-pal2: Configuring for [Linux], and [1]

```
geunsik@FCS-Server1:~/dotnet/coreclr$ time ROOTFS_DIR=/work/dotnet/rootfs-coreclr/arm  ./build.sh arm debug clean  cross skipmscorlib | grep INFO
loading initial cache file /home/geunsik/dotnet/coreclr/cross/arm/tryrun.cmake
Detected Linux ARM
-- INFO-pal2: Detected Linux (from ./src/pal/src/configure.cmake) 
-- INFO-pal2: Configuring for [Linux], and [1 ]
CMake Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:
   HAVE_CLOCK_MONOTONIC_COARSE_EXITCODE (advanced)
For details see /home/geunsik/dotnet/coreclr/bin/obj/Linux.arm.Debug/TryRunResults.cmake
-- INFO-pal1: Detected Linux (from src/pal/src/CMakeLists.txt)
-- INFO-pal1: Configuring for [Linux], and [1]

real	0m7.308s
user	0m5.231s
sys	0m2.135s
geunsik@FCS-Server1:~/dotnet/coreclr$ 
```

**Comment by janvorli** [comment]
@leemgs I don't understand what you are trying to fix. The `__LINUX__` is supposed to be C/C++ definition with value set to 1, not a cmake variable. Why have you added the cmake variable? Regarding the issue #3479, the correct symbol to use in the configure.cmake was really the `__linux__` as you had it before, as I've realized from the response of @myungjoo

**Comment by leemgs** [comment]
@janvorli, Yes. The issue https://github.com/dotnet/coreclr/pull/3479 is completed. As a next step, I am checking normal operation of the inherited __LINUX__ value between __LINUX__ in ./coreclr/src/pal/* and "add_definitions(-D__LINUX__=1)" in ./coreclr/CMakeLists.txt


```
invain@E120:/work/dotnet/dotnet-arm-32bit/coreclr$ 
invain@E120:/work/dotnet/dotnet-arm-32bit/coreclr$ grep -R "__LINUX__" ./*
./CMakeLists.txt:    add_definitions(-D__LINUX__=1)
./src/inc/pedecoder.h:#elif defined(__LINUX__)
./src/pal/src/CMakeLists.txt:  add_definitions(-D__LINUX__=1)
./src/pal/src/loader/module.cpp:#if defined(__LINUX__)
./src/pal/src/include/pal/thread.hpp:#if defined(__LINUX__)
./src/pal/src/init/pal.cpp:#if defined(__LINUX__)
./src/pal/src/misc/tracepointprovider.cpp:#if defined(__LINUX__)
./src/pal/src/misc/sysinfo.cpp:#elif defined(__sun__) || defined(_AIX) || defined(__hppa__) || ( defined (_IA64_) && defined (_HPUX_) ) || defined(__LINUX__)
./src/pal/src/misc/sysinfo.cpp:#if HAVE_SYSCONF && defined(__LINUX__)
./src/pal/src/exception/seh-unwind.cpp:#ifndef __LINUX__
./src/pal/src/exception/seh-unwind.cpp:#endif // !__LINUX__       
./src/pal/src/exception/seh-unwind.cpp:#ifdef __LINUX__
./src/pal/src/exception/seh-unwind.cpp:#endif // __LINUX__    
./src/pal/src/exception/seh.cpp:#elif __LINUX__ || defined(__FreeBSD__) || defined(__NetBSD__)
./src/pal/src/exception/seh.cpp:#elif __LINUX__ || defined(__FreeBSD__) || defined(__NetBSD__)
./src/pal/tests/CMakeLists.txt:add_definitions(-D__LINUX__=1)
./src/pal/inc/pal.h:#elif defined(__LINUX__) && defined(_ARM_)
./src/pal/inc/pal.h:#elif defined(__LINUX__) && defined(_ARM64_)
./src/pal/inc/pal.h:#elif defined(__LINUX__) && defined(__x86_64__)
./src/pal/inc/pal_mstypes.h:#if defined(__APPLE_CC__) || defined(__LINUX__)
./src/pal/inc/pal_mstypes.h:#if defined(__LINUX__) 
./src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp:#if defined(__LINUX__)
./src/coreclr/hosts/unixcoreruncommon/coreruncommon.cpp:#if defined(__LINUX__)
invain@E120:/work/dotnet/dotnet-arm-32bit/coreclr$ 

```

**Comment by jkotas** [comment]
According to https://sourceforge.net/p/predef/wiki/OperatingSystems/, `__linux__` is the right way to detect Linux.

We should replace all uses of `__LINUX__` with the lower case one, and delete our own uppercase definition. Are you going to submit the PR?

**Comment by leemgs** [comment]
@jkotas, Yes. I will send PR after build and running test.

**Comment by leemgs** [comment]
@jkotas, I have sent PR https://github.com/dotnet/coreclr/pull/3528. 

Below is generated binary files. 


```
invain@E120:/work/dotnet/arm-32bit/coreclr$ ls -al ./bin/Product/Linux.arm.Release/
합계 223372
drwxrwxr-x 5 invain invain     4096  3월  5 11:51 .
drwxrwxr-x 3 invain invain     4096  3월  5 11:22 ..
-rw-r--r-- 1 invain invain   351755  3월  5 11:23 System.Globalization.Native.so
-rwxr-xr-x 1 invain invain   125077  3월  5 11:49 coreconsole
-rwxr-xr-x 1 invain invain   123390  3월  5 11:49 corerun
-rwxr-xr-x 1 invain invain 50492251  3월  5 11:44 crossgen
drwxrwxr-x 2 invain invain     4096  3월  5 11:51 gcinfo
-rwxr-xr-x 1 invain invain  5187216  3월  5 11:44 ilasm
-rwxr-xr-x 1 invain invain  4767397  3월  5 11:44 ildasm
drwxrwxr-x 2 invain invain     4096  3월  5 11:51 inc
drwxrwxr-x 2 invain invain     4096  3월  5 11:51 lib
-rw-r--r-- 1 invain invain 79484525  3월  5 11:49 libcoreclr.so
-rw-r--r-- 1 invain invain  3307594  3월  5 11:43 libdbgshim.so
-rw-r--r-- 1 invain invain 37485587  3월  5 11:50 libmscordaccore.so
-rw-r--r-- 1 invain invain 11683533  3월  5 11:50 libmscordbi.so
-rw-r--r-- 1 invain invain 16467912  3월  5 11:44 libprotojit.so
-rw-r--r-- 1 invain invain 16413060  3월  5 11:45 libryujit.so
-rw-r--r-- 1 invain invain  2346962  3월  5 11:51 libsos.so
-rw-r--r-- 1 invain invain   387585  3월  5 11:51 libsosplugin.so
-rw-r--r-- 1 invain invain    61414  3월  5 11:09 sosdocsunix.txt
invain@E120:/work/dotnet/arm-32bit/coreclr$ file ./bin/Product/Linux.arm.Release/corerun 
./bin/Product/Linux.arm.Release/corerun: ELF 32-bit LSB  executable, ARM, EABI5 version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=7ddb06ab040d01bc5a4812b785ce16882cda4079, not stripped
invain@E120:/work/dotnet/arm-32bit/coreclr$ file ./bin/Product/Linux.arm.Release/libcoreclr.so 
./bin/Product/Linux.arm.Release/libcoreclr.so: ELF 32-bit LSB  shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, BuildID[sha1]=128faf51d71871ba8a378b871e3aa97f64cd36df, not stripped
invain@E120:/work/dotnet/arm-32bit/coreclr$ time ROOTFS_DIR=/work/dotnet/rootfs-coreclr/arm  ./build.sh arm release clean cross skiptests
```

**Comment by jkotas** [comment]
Fixed by #3528
